Refactor uart

ledvinap opened the pull request:
- **serial - refactor serial resource handling**
    - port related resources (tx/rx/inverted,dma)
      - are stored sequentially, with space in place of ports that are not enabled (RESOURCE_UART_COUNT + RESOURCE_LPUART_COUNT + RESOURCE_SOFTSERIAL_COUNT)
      - RESOURCE_UART_OFFSET, RESOURCE_LPUART_OFFSET, RESOURCE_SOFTSERIAL_OFFSET
      - resource entries are pointing into this array (UART, LPUART, SOFTSERIAL)
        - both pins and DMA
        - inverter is supproted only for UART + LPUART (makes sense only for UART, to be removed)
      - softSerialPinConfig is removed, it is no longer necessary
    - serialResourceIndex(identifier) is used universally to get correct index into resource array
    - unified handling of resources where possible
    - serialOwnerTxRx() + serialOwnerIndex() are used for displaying resources correctly to user.
    - serialType(identifier) implemented
      - serialOwnerTxRx / serialOwnerIndex are trivial with it
      - large switch() statemens are greatly simplified
    

- **serial - merge code duplicated in all UART implementations**
    - drivers/serial_uart_hw.c contains merged serialUART code.
      Code did not match exactly. Obvious cases are fixed, more complicated use #ifs
    - pin inversion unified
    - uartOpen is refactored a bit
    

- **serial - refactor uartDevice**
    - use 'compressed' enum from uartDeviceIdx_e. Only enabled ports are in this enum.
    - uartDeviceIdx directly indexes uartDevice (no search necessary, no wasted space)
    - use `serialPortIdentifier_e identifier;` for uartHardware
    - for DMA remap, define only entries for enabled ports (uartDeviceIdx_e does not exist disabled port)
    

- **serial - apply changes to inverter**
    New implementation is trivial
    

- **serial - HAL - rxIrq, txIrq replaces by irqn**
    There is only one IRQ for serial
    

- **serial - serial_post.h**
    Generated code to normalize target configuration.
    jinja2 template is used to generate header file. Port handling is unified a lot.
    
    `SERIAL_<type><n>_USED 0/1` - always defined, value depends on target configuration
    `SERIAL_<type>_MASK`        - bitmask of used ports or given type. <port>1 is BIT(0)
    `SERIAL_<type>_COUNT`       - number of enabled ports of given type
    `SERIAL_<type>_MAX`         - <index of highest used port> + 1, 0 when no port is enabled
    

- **targets - remove automatically calculated valued from configs**
    serial_post.h generated them
    

- **serial - change LPUART to UART_LP1 in places**
    LPUART is mostly handled as another UART port, this change reflects it
    

- **serial - use ARRAYLEN / ARRAYEND in some places**
    replaces constant that may not match actual array size
    
- **serial - adapt softserial to changes**
    

- **softserial - fix serialTimerConfigureTimebase**
    

- **serial - simpler port names**
    Is this useful ?

- **serial - no port count for SITL necessary**


- **serial - fix unittests**
    include serial_post.h, some ports are defined, so normalization will
    work fine


- **timers - remove obsolete defines from SITL and unittests**


- **Merge 'upstream/master' into refactor-uart**
    Conflicts:
        src/main/drivers/mcu/at32/serial_uart_at32f43x.c
        src/main/drivers/mcu/stm32/serial_uart_stm32f4xx.c
        src/main/drivers/mcu/stm32/serial_uart_stm32g4xx.c

    Changes applied to serial_uart_hw.c


- **serial - fix serialInit bug**
    serialPortToDisable was preventing further tests


- **serial - fix possible NULL pointer dereference**


- **serial - move serialResourceIndex to drivers**


- **serial - refactor pp/od and pulldown/pullup/nopull**
    Centralize serial options handling, decouple it from MCU type

    Move some code into new drivers/serial_impl.c


- **serial - make port.identifier valid early in port initialization**


- **serial - fix unittest**
    Code is moved around a bit to make unittests implementation easier


- **serial - bug - fix off-by-one error**
    uart1/softserial1/lpuart was not working properly because of this


- **utils - add popcount functions**
    Wrap __builtin_popcount for easier use. Refactor existing code using
    it. Update code using BITCOUNT macro in dynamic context (BITCOUNT is
    for compile-time use only)


- **serial - inverter - simplify code**


- **pinio - prevent array access overflow**


- **serial - refactor SERIAL_BIDIR_***
    SERIAL_BIDIR_OD / SERIAL_BIDIR_PP
    SERIAL_PULL_DEFAULT / SERIAL_PULL_NONE / SERIAL_PULL_PD


- **serial - simplify code**
    minor refactoring
    - cleaner AVOID_UARTx_FOR_PWM_PPM (unused anyway)
    - serialFindPortConfiguration* - remove code duplication
    - use serialType in smartaudio


- **serial - add port names**
    - const string is assiociated with each compiled-in port (easy to pass around)
    - findSerialPortByName


- **cli - improve serialpassthrough**
    - support port options (in current mode argument)
    - support port names (`serialpassthrough uart1`)
    - improve error handling; check more parse errors


- **cli - resources - minor refactor**
    - prevent SERIAL_TX_NONSENSE when parsing resource type
    - fix possible NULL pointer dereference
    - store resource tag only after checking all conditions
    - slightly better error reporting
    - `resource serial_tx 1` will print resource assignment


- **serial - rename pulldown to SERIAL_PULL_SMARTAUDIO**
    Make sure nobody uses it by
    mistake. SERIAL_PULL_SMARTAUDIO_SUPER_DANGEROUS_USE_AT_YOUR_OWN_RISK_THIS_WILL_BREAK_EVERYTHING_AND_KILL_YOUR_PET_FISH
    would be better choice, but try shorter version first.


- **smartaudio - minor refactor/improvement**
    - Fix softserial on AT32
    - make it easier to handle SA on per-cpu basis


- **UART - move AF handling before MCU dependent code**


- **UART - adapt APM32 / APM32F4 code**
    - Modeled after F4 where possible
    - some code moved from APM32 to APM32F4, possibly not necessary, but
       it improves similarity with STM32F4


- **UART - APM32 - remove per-pin AF**


- **UART - APM32 - bugfix - fix pinswap #if conditions**


- **UART - apply some improvements from APM32 to STM32**


- **UART - add todo for F4 serial speed**


- **UART - refactor support for USE_SMARTAUDIO_NOPULLDOWN**


- **UART - move code into enableRxIrq**
    TODO: split into mcu-specific filer (but in sepatate PR)


- **UART - add UART_TRAIT_xx**
    makes #if test easier and more consistent
    more traits shall follow


- **UART - use tables instead of switch**


- **UART - smartaudio**
    minor style change + better comments


- **UART - unify mspPorts iteration**


- **UART - use OwnerIndex for inverter**
    - 1 based
    - only UART has inversion


- **UART - refactor serial_resource a bit**
    Single table + helper function. New table is easier to read


- **UART - remove LPUART inverter**
    LPUART always supports internal inversion, code was incomplete and
    unused

- **UART - enable UART module RCC clk before configuring DMA**
    Does not change anything, UART RCCs are enabled unconditionally

ledvinap commented:
A bit bigger PR .. but it removes ~930 lines of code (generated serial_post.h is 530 lines)

I can cleanup it a bit or remove more controversial parts.

haslinghuis commented on src/main/cli/cli.c:
```diff
@@ -5022,20 +5022,21 @@ const cliResourceValue_t resourceTable[] = {
-#ifdef USE_LPUART1
-    DEFA( OWNER_LPUART_TX,     PG_SERIAL_PIN_CONFIG, serialPinConfig_t, ioTagTx[SERIAL_UART_COUNT], SERIAL_LPUART_COUNT ),
-    DEFA( OWNER_LPUART_RX,     PG_SERIAL_PIN_CONFIG, serialPinConfig_t, ioTagRx[SERIAL_UART_COUNT], SERIAL_LPUART_COUNT ),
+#if defined(USE_LPUART)
```
```suggestion
#if defined(USE_LPUART1)
```

haslinghuis commented on src/main/drivers/serial_pinconfig.c:
```diff
@@ -34,159 +34,7 @@
-#endif
-
-#if defined(USE_UART) || defined(USE_SOFTSERIAL)
+#if defined(USE_UART) || defined(USE_LPUART) || defined(USE_SOFTSERIAL)
```
```suggestion
#if defined(USE_UART) || defined(USE_LPUART1) || defined(USE_SOFTSERIAL)
```

haslinghuis replied:
Only have one LPUART at the moment - but we have ID's 40-49 reserved for LPUART 1-10 so maybe we go with `USE_LPUART` instead to be consistent.

SteveCEvans commented:
Impressive amount of work. Have you tested DMA operation? This is code we don’t use and so isn’t tested at the moment, but in future and with more DMA channels this may be resurrected.

ledvinap commented:
@SteveCEvans : DMA should be unchanged (same enums, with possibly different values). I'll double-check it.

haslinghuis commented:
@ledvinap please rebase

ledvinap commented:
Still waiting for HW to test on, don't merge yet. But I'm not aware of problems with this code.

nerdCopter commented:
will need merge master or rebase

ctzsnooze commented:
Can an ordinary end-user like me test?  Is it useful to flash this PR and just check that GPS, Rx etc still work?

ledvinap commented:
I'll rebase it soon.

There may be some problems with exotic (or magic - SA) hardware, if I unified UART pin/AF/pullup/monitor handling from wrong version.

nerdCopter commented:
not sure how to test on unsoldered FC, but settings get saved, etc.
i suppose this needs flight testing with peripherals.

![image](https://github.com/user-attachments/assets/af9dd036-1bf6-46a1-b339-23c6313182e8)
```
# version
# Betaflight / AT32F435M (A435) 4.6.0 Aug  8 2024 / 18:11:06 (e46e93286) MSP API: 1.47
# config rev: 67eee40
# board: manufacturer_id: AIRB, board_name: AIRBOTSUPERF4

# status
MCU AT32F435 Clock=288MHz, Vref=3.30V, Core temp=50degC
Stack size: 2048, Stack address: 0x2002fff0
Configuration: CONFIGURED, size: 3906, max available: 16384
Devices detected: SPI:1, I2C:1
Gyros detected: gyro 1 locked dma
GYRO=ICM42688P, ACC=ICM42688P, BARO=DPS310
OSD: MAX7456 (30 x 13)
BUILD KEY: ae3e164536daafd9c242807a23c9f33a (4.6.0-dev)
System Uptime: 149 seconds, Current Time: 2024-08-08T18:28:41.603+00:00
CPU:12%, cycle time: 124, GYRO rate: 8064, RX rate: 15, System rate: 9
Voltage: 29 * 0.01V (0S battery - NOT PRESENT)
I2C Errors: 0
FLASH: JEDEC ID=0x00ef4018 16M
GPS: NOT CONNECTED, UART7 115200 (set to 57600), NOT CONFIGURED, version =  unknown
Arming disable flags: RXLOSS CLI MSP
```
![image](https://github.com/user-attachments/assets/a2acf04a-57cb-4672-8078-6611f22eac82)

haslinghuis commented:
Tested BETAFPVF4SX1280 which uses SA over softserial - it does not work with this PR. (Softserial is showing up). Do not have access to SA capable hardware for testing UART.

ledvinap commented on src/main/drivers/serial_uart.c:
```diff
@@ -112,21 +108,100 @@ UART_BUFFERS(10);
+        return uartMap[identifier] ? uartMap[identifier] - 1 : UARTDEV_INVALID;
+    }
+#else
+    {
+        int idx = identifier - SERIAL_PORT_USART1;
+        if (idx >= 0 && idx < SERIAL_UART_MAX) {
+            if (BIT(idx) & SERIAL_UART_MASK) {
+                // retrun number of enabled uart ports smaller than idx
+                return __builtin_popcount((BIT(idx) - 1) & SERIAL_UART_MASK);  // TODO - pocount in utils.h
+            } else {
+                return UARTDEV_INVALID;
+            }
+        }
+    }
```
What about this code? It is not enabled now, but I think it is production ready

ledvinap commented:
I'll add few commits soon...

haslinghuis commented:
**BETAFPVF4SX1280**: SmartAudio 2.0 over softserial not working.

**AIRBOTG4AIO**: LPUART is present, but there is a nonexisting SOFTSERIAL port available ? (feature SOFTSERIAL is disabled)

haslinghuis commented:
@ledvinap BETAFPVF4SX1280 works in master.

With AIRBOTG4AIO it shows SOFTSERIAL port in ports tab (while there are no resources).

AIRBOTG4AIO master: fa9edde3-e68d-4811-908d-b37501c9cc0b
AIRBOTG4AIO PR: b2868f00-6f1a-42f7-a1c5-fbfb238f0d8c

ledvinap commented:
@haslinghuis : AIRBOTG4AIO : Any chance that old config was left on board? Resource layout is changed in this PR, it may be causing problems (I need to look at versions)

haslinghuis commented:
Always flash with full chip erase

haslinghuis commented:
@ledvinap No - flashing with full chip erase by default

ledvinap commented:
@haslinghuis : my stupid mistake. Now first ports shall work again. 

SmartAudio still needs some work

haslinghuis commented:
![image](https://github.com/user-attachments/assets/cee4d01c-d53a-42bf-999f-34da6a6ebde2)

AIRBOTG4AIO ports looks okay now !!!

JyeSmith commented:
Nothing has changed with OVTX in a long time.  Anything on the BF side of things that has changed to prevent flashing?

nerdCopter commented:
`18d393c`
`BETAFPVF4SX1280` (https://betafpv.com/products/f4-1s-5a-aio-brushless-flight-controller)
SmartAudio over SoftSerial worked: https://youtu.be/FheoakSzgLw

haslinghuis commented:
Tested on SPBE F7V3 with TBS Unify Pro 5g8 v3 (SA 2.0) and GEP M1025-MI.

@ledvinap please rebase. Think this is ready to merge or is there anything left to do ?

nerdCopter reviewed (approved):
- approving for workflow. suggest don't merge until @ledvinap replies that it is good to go.

haslinghuis commented:
@ledvinap this will be hard to rebase :(

ledvinap commented:
@haslinghuis : Not that bad, but I need to read it again, to check APM32 code

haslinghuis commented on src/main/drivers/serial_uart_hw.c:
```diff
@@ -0,0 +1,200 @@
+                // TODO: no need for pullup on TX only pin
+                IOConfigGPIOAF(txIO, IOCFG_AF_PP_UP, txAf);
+#else
+                IOConfigGPIOAF(txIO, IOCFG_AF_PP, txAaf);
```
```suggestion
                IOConfigGPIOAF(txIO, IOCFG_AF_PP, txAf);
```

haslinghuis reviewed (commented):
This fixes the build

haslinghuis commented:
Conflict is the SA hack. Suggest making it independent on softserial as the issue also reported using an UART.

haslinghuis commented:
@ledvinap tested. Removed draft label. Should be ready to merge now.

ledvinap commented:
Can someone test APM32? Or at least review code related to it?

haslinghuis commented:
@ledvinap There a no official APM32 target available. Have a STELLARF4 which uses Geehy APM32F05 but it just works with `STM32F405` defined. 
On master or this PR it does not work when changing to `#define FC_TARGET_MCU APM32F405`

ledvinap commented:
Don't merge it yet, I'll verify the diff once more ...

haslinghuis commented:
Update: board needs a cold boot after flashing on master or PR. But APM32 seems to work

```
# version
# Betaflight / APM32F405 (A405) 4.6.0 Oct 27 2024 / 20:20:27 (244d40bdd) MSP API: 1.47
# board: manufacturer_id: STBE, board_name: STELLARF4

# status
MCU APM32F405 Clock=168MHz (PLLP-HSI), Vref=3.28V, Core temp=30degC
Stack size: 2048, Stack address: 0x1000fff0
Configuration: CONFIGURED, size: 3984, max available: 16384
Devices detected: SPI:1, I2C:1
Gyros detected: gyro 1 locked dma
GYRO=BMI270, ACC=BMI270, BARO=BMP280
OSD: MAX7456 (30 x 13)
System Uptime: 220 seconds, Current Time: 2024-10-27T22:32:56.037+00:00
CPU:17%, cycle time: 374, GYRO rate: 2673, RX rate: 0, System rate: 9
Voltage: 0 * 0.01V (0S battery - NOT PRESENT)
I2C Errors: 0
FLASH: JEDEC ID=0x00ef4018 16M
GPS: NOT ENABLED
Arming disable flags: RXLOSS CLI MSP

# tasks
Task list             rate/hz  max/us  avg/us maxload avgload  total/ms   late    run reqd/us
00 - (         SYSTEM)     10       1       0    0.0%    0.0%         2      0     12       4
01 - (         SYSTEM)    982       6       0    0.5%    0.0%       234      0   1084       3
02 - (           GYRO)   3182       6       3    1.9%    1.0%      3947      0   3522       0
03 - (         FILTER)   3182      27      19    8.5%    6.1%     24238      0   3522       0
04 - (            PID)   3182      50      39   15.9%   12.6%     50719      0   3522       0
05 - (            ACC)    792      20       3    1.5%    0.2%       945      1    881      14
06 - (       ATTITUDE)     99      12       8    0.1%    0.0%       335      0    111      11
07 - (             RX)     20      24      16    0.0%    0.0%       409      0     69      23
08 - (         SERIAL)     99   21926       2  217.0%    0.0%      4732      0    111      31
09 - (       DISPATCH)    982       6       0    0.5%    0.0%       259      6   1084       4
10 - (BATTERY_VOLTAGE)     50       4       1    0.0%    0.0%        33      0     56       6
11 - (BATTERY_CURRENT)     50       3       0    0.0%    0.0%        15      0     56       4
12 - ( BATTERY_ALERTS)      5       2       1    0.0%    0.0%         3      0      6      14
13 - (         BEEPER)     99       1       0    0.0%    0.0%        32      0    111       2
16 - (           BARO)     40      16       7    0.0%    0.0%       353      0    135      12
17 - (       ALTITUDE)     99       3       2    0.0%    0.0%       102      0    111      15
19 - (       LEDSTRIP)     60       0       0    0.0%    0.0%        20      1     67       0
20 - (            OSD)     12      41      20    0.0%    0.0%      1699      2    302      39
22 - (            CMS)     20       2       1    0.0%    0.0%        11      0     23       3
23 - (        VTXCTRL)      5       1       0    0.0%    0.0%         1      0      6       8
25 - (    ADCINTERNAL)      1       4       3    0.0%    0.0%         1      0      1       7
27 - (SPEED_NEGOTIATION)     99       7       1    0.0%    0.0%        57      0    110      15
RX Check Function                   9       3                       518
Total (excluding SERIAL)                                19.9%
```

ledvinap commented:
@haslinghuis : Thanks!

haslinghuis commented on src/main/drivers/serial_resource.h:
```diff
@@ -0,0 +1,39 @@
+
+/*
+  #defines used for serial port resource access (pin/dma/inversion)
+  target/serial_post.h normalizes enabled port definitions, values are just remaned here
```
```suggestion
  target/serial_post.h normalizes enabled port definitions, values are just remained here
```

ledvinap replied:
Thanks!
Actually  `renamed` is correct.

haslinghuis commented on src/main/io/serial_resource.c:
```diff
@@ -0,0 +1,134 @@
+        }
+    }
+#endif
+#ifdef USE_LPUART
```
Note `USE_LPUART` and `USE_LPUART1` in this PR.

ledvinap replied:
All ports are normalized consistently. USE_LPUART1 means that LPUART1 is enabled, USE_LPUART means that any (but there is only LPUART1 now) is enabled.
I think it is used correctly.

haslinghuis replied:
![Screenshot from 2024-10-30 18-27-54](https://github.com/user-attachments/assets/d71e2f7a-aa17-4694-9eab-d7a1f8b9bb44)

Please have a look why LPUART shows up for MATEKH743.

ledvinap replied:
fixed

haslinghuis reviewed (commented):
One issue found: tested PR on MATEKH743 it exposes non existing LPUART.

ledvinap commented:
> One issue found: tested PR on MATEKH743 it exposes non existing LPUART.

I don't know how to handle this - H743 (https://github.com/betaflight/betaflight/pull/13585/files#diff-ee1dd7bc6fccc142470bba593a8a432a8c5cc9dd31305712026418129adae5a2R46) target.h enables LPUART1. 
This PR should not change that logic (LPUART1 does exist in H743, on all packages. It is different problem that it is not routed to pad)

I can imagine some automatic mechanism to enable only peripherals that connected on FC (TARGET_IO_PORTx was doing this for IO, but it can be reused for serial, with nontrivial, but manageable amount of magic).

ledvinap commented:
@haslinghuis : You you post support ID, please? I don't understand UART5  - it should be available ...

haslinghuis commented:
USART5 in [config](https://github.com/betaflight/config/blob/master/configs/MATEKH743/config.h) is not broken out on [MATEKH743](https://www.mateksys.com/?portfolio=h743-slim#tab-id-7): 

**master**:  [7812a060-8746-496a-aa01-713b56c3bf1e](https://build.betaflight.com/api/support/7812a060-8746-496a-aa01-713b56c3bf1e)

![image](https://github.com/user-attachments/assets/285ee182-ecc1-48b5-9c5a-40985d1dd9e3)

#**13585**: [219265f6-f4ac-459d-ac2f-3a6626ec3755](https://build.betaflight.com/api/support/219265f6-f4ac-459d-ac2f-3a6626ec3755)

![image](https://github.com/user-attachments/assets/333a2c27-214d-458e-a91a-565bf86f017e)

haslinghuis commented:
@ledvinap there is no problem with UART5

USART5 in [config](https://github.com/betaflight/config/blob/master/configs/MATEKH743/config.h) is not broken out on [MATEKH743](https://www.mateksys.com/?portfolio=h743-slim#tab-id-7): 

**master**:  [7812a060-8746-496a-aa01-713b56c3bf1e](https://build.betaflight.com/api/support/7812a060-8746-496a-aa01-713b56c3bf1e)

![image](https://github.com/user-attachments/assets/285ee182-ecc1-48b5-9c5a-40985d1dd9e3)

#**13585**: [219265f6-f4ac-459d-ac2f-3a6626ec3755](https://build.betaflight.com/api/support/219265f6-f4ac-459d-ac2f-3a6626ec3755)

![image](https://github.com/user-attachments/assets/333a2c27-214d-458e-a91a-565bf86f017e)

ledvinap commented:
@haslinghuis: Fixed. I broke it during simple serialResourceIndex refactoring.

Thanks for spotting it, great work!

Can you post support again, please?

haslinghuis reviewed (approved):
6d9eabf6-eaf3-41f1-b578-ed933e61bd34

haslinghuis commented:
Sorry we are not done yet.

[AIRBOTG4AIO](https://github.com/betaflight/config/tree/master/configs/AIRBOTG4AIO) no longer showing up after flashing PR.

On master: [d0af62ca-6fcb-4e2e-8b2b-007d023751b6](https://build.betaflight.com/api/support/d0af62ca-6fcb-4e2e-8b2b-007d023751b6)

ledvinap commented:
@haslinghuis : Thanks for testing. 
The behavior is really strange. Gcc is optimizing quite aggressively, but I don't see what can cause problems.
Can you try veriant with const?

BTW: gcc ARM EABI has -fpack-enums enabled. It may be related to this problem and it may be causing other problems

haslinghuis commented:
@ledvinap Thanks. With latest commit https://github.com/betaflight/betaflight/pull/13585/commits/2b51aee9f5119f435c09504a4a7c4a5bfbe227d4 both boards work as expected !

ledvinap commented:
@haslinghuis : Sorry, it is `-fshort-enums`

ledvinap commented:
@haslinghuis : Can you compile  AIRBOTG4AIO with 
```
make -k AIRBOTG4AIO EXTRA_LD_FLAGS="-save-temps -fverbose-asm"
```
And upload .s filed from obj/main/betaflight_STM32G47X_AIRBOTG4AIO.elf.ltrans*.ltrans.s 

And can you try different GCC version?



This is bad. The function is IMO fine (generated assembly looks ok), adding 'const' only allows gcc to inline it more aggressively.
I don't see any undefined behavior or similar problem.

So the problem may be caused by stray memory access, with this only changing memory layout a bit.

haslinghuis commented:
@ledvinap Using gcc 11.4.0 (should be same as CI version)

`make -k AIRBOTG4AIO EXTRA_LD_FLAGS="-save-temps -fverbose-asm"` produces [airbotg4aio.zip](https://github.com/user-attachments/files/17603171/airbotg4aio.zip)

ledvinap commented:
@haslinghuis : Sorry, the version that does not work, please

haslinghuis commented:
Now it does not fail even without the `const` on AIRBOTG4AIO. Will flash back to master and reapply PR locally. This is weird.

haslinghuis commented:
@ledvinap this build is without constant (serial_resource.c:118) but now it does not fail. Using same command from log and have tested multiple times before. Have also flashed PR using cloud build in previous state but works in current state of PR.  

`make -k AIRBOTG4AIO EXTRA_LD_FLAGS="-save-temps -fverbose-asm"`
[airbotg4-without-constant.zip](https://github.com/user-attachments/files/17603569/airbotg4-without-constant.zip)

haslinghuis commented:
@ledvinap did you find anything disturbing or is this ready to merge ?

ledvinap commented:
@haslinghuis : I checked the diff again, I don't see any problem. Generated ASM for serialResourceIndex was correct, 

I refactored serial_resource a bit, mainly improving code readability

Added few changes after reading diff again. Should be cosmetics only (IIRC only actual change is correct INVERTER index in `resource`).

ledvinap commented:
@haslinghuis : Can you please test one more, just to be sure?

haslinghuis commented:
Tested with MATEKH743, AIRBOTG4AIO and SPEEDYBEEF7V3

haslinghuis commented:
@ledvinap Tested with MATEKH743, AIRBOTG4AIO and SPEEDYBEEF7V3

nerdCopter commented:
i cannot connect to my foxxerF405 with this merge.
`a659189bf` worked, `246d04dc5` does not.

`a659189bf`
```
$ sudo dmesg | tail
[92745.135903] usb 2-9.4.2: Manufacturer: Betaflight
[92745.135904] usb 2-9.4.2: SerialNumber: 0x8000000
[92745.143557] cdc_acm 2-9.4.2:1.0: ttyACM0: USB ACM device
[92802.645923] usb 2-9.4.2: USB disconnect, device number 72
[92802.873292] usb 2-9.4.2: new full-speed USB device number 73 using xhci_hcd
[92802.978628] usb 2-9.4.2: New USB device found, idVendor=0483, idProduct=df11, bcdDevice=22.00
[92802.978633] usb 2-9.4.2: New USB device strings: Mfr=1, Product=2, SerialNumber=3
[92802.978634] usb 2-9.4.2: Product: STM32  BOOTLOADER
[92802.978635] usb 2-9.4.2: Manufacturer: STMicroelectronics
[92802.978636] usb 2-9.4.2: SerialNumber: 207234A14E48
```

`246d04dc5`
```
$ sudo dmesg | tail
[92863.433620] usb 2-9.4.2: Product: STM32  BOOTLOADER
[92863.433622] usb 2-9.4.2: Manufacturer: STMicroelectronics
[92863.433623] usb 2-9.4.2: SerialNumber: 207234A14E48
[92906.837967] usb 2-9.4.2: USB disconnect, device number 75
[92908.859785] usb 2-9.4.2: new full-speed USB device number 76 using xhci_hcd
[92923.703600] usb 2-9.4.2: new full-speed USB device number 77 using xhci_hcd
[92939.171382] usb 2-9.4.2: device descriptor read/64, error -110
[92954.787151] usb 2-9.4.2: device descriptor read/64, error -110
[92954.975146] usb 2-9.4.2: new full-speed USB device number 78 using xhci_hcd
[92972.082896] usb 2-9.4.2: new full-speed USB device number 79 using xhci_hcd
```

haslinghuis commented:
Tested on BETAFPVF4SX1280, TUNERCF405 and STELLARF4 without issues
This STELLARF4 is using APM32 Geehy (configured as F405) and needs a cold boot after flashing as was the case before this PR.

rabbitambulance-ra commented:
> i cannot connect to my foxeerF405 with this merge. i had not tested more recent changes.

@ledvinap, this PR unfortunately also breaks the HGLRC F722 mini (older one with the MPU6000), target HGLR/HGLRCF722()STM32F7X2). 

Flashing latest 4.6.0-dev [latest] results in the board not booting, and no USB device showing up after flash. Only thing possible is holding DFU button and re-flashing.

Flashing any 4.5.x (including 4.5.2-dev) or 4.4.x, _as well as_ the PR right before this one ("Yaw type rudder/diff_thrust for TPA calculations...") allows the HGLRC F722 mini to once again boot correctly. 

Tested and confirmed across two different brand new HGLRC F722 mini FCs, so definitely not a hardware issue.

nerdCopter commented:
`AIRBOTG4AIO` `246d04dc5` succeeds.
`PYRODRONEF7` `246d04dc5` succeeds.
`SKYSTARSF405AIO` `246d04dc5` succeeds.

`AIRBOTSUPERF4` AT32 `246d04dc5` fails, `a659189bf` succeeds, unsoldered.
```
$ sudo dmesg | tail
[...]
[99193.148514] usb 2-9.4.2: new full-speed USB device number 19 using xhci_hcd
[99198.316430] usb 2-9.4.2: device descriptor read/64, error -32
[99204.156293] usb 2-9.4.2: new full-speed USB device number 20 using xhci_hcd
[99219.532061] usb 2-9.4.2: device descriptor read/64, error -110
```

ledvinap commented:
Working on it, can reproduce on AIRBOTSUPERF4 now.
It seems to be USB related - VCP does not work even without USE_UART

haslinghuis commented:
Confirm AIRBOTSUPERF4 failing

```
[23158.097027] usb 1-1: device not accepting address 37, error -62
[23158.097189] usb 1-1: WARN: invalid context state for evaluate context command.
[23158.211846] usb 1-1: new full-speed USB device number 38 using xhci_hcd
[23163.520927] xhci_hcd 0000:00:14.0: Timeout while waiting for setup device command
[23169.153096] xhci_hcd 0000:00:14.0: Timeout while waiting for setup device command
[23169.360990] usb 1-1: device not accepting address 38, error -62
[23169.361144] usb 1-1: WARN: invalid context state for evaluate context command.
[23169.361238] usb usb1-port1: unable to enumerate USB device
[23402.327443] usb 1-1: new full-speed USB device number 39 using xhci_hcd
[23418.100522] usb 1-1: device descriptor read/64, error -110
```

ledvinap commented:
@betaflight: Anyone with debugger connected to failing target? It's quite possible it's hardfaulting, it would be great to know details ...

ledvinap commented:
I probably got it, fix soon

ledvinap commented:
Please test  https://github.com/betaflight/betaflight/pull/14007
But I doubt that it does fix the problem, it is probably only masking it ...

rabbitambulance-ra commented:
> Please test #14007 But I doubt that it does fix the problem, it is probably only masking it ...

This did indeed fix it for me on the HGLRC F722 Mini! Thanks for the quick turnaround, @ledvinap! 
![image](https://github.com/user-attachments/assets/96bdacb7-7bbe-46fb-a486-a035bcf42d66)

ledvinap commented:
Fixed in #14007
